### PR TITLE
Multilabel

### LINKED ISSUE #16: Use Efficientnetv2 from tensorflow

### PR DESCRIPTION
**Pull request by mitul23**
adding a new class to the testing script and all the utility to handle multilabel image classification.

### CODE REVIEW

File: decavision/model_testing/testing.py [hunk 1]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+
+
+class ModelTesterMultilabel:
+    """
+    Class to use a trained multilabel image classification model on some images. Using this
+    when working with a TPU disables eager execution.
+
+    Arguments:
+        model (str): path to trained model
+        json_file (str): path to json file containing image ids and their associated labels
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: I would actually use the json as an argument in the methods that need it directly. This way people could use the same class and call a method like `classify_images` for different datasets.

File: decavision/model_testing/testing.py [hunk 2]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+        self.values = values
+        self.json_file = json_file
+        self.categories = categories
+        print("Model name: ",self.model.name)
+
+
+    def _load_dataset(self, path):
+        """
+        Load dataset into a keras generator. Images must be contained in single
+        folder for each class. A dataframe is required with 2 columns: original
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: At many places in the docstrings it says that the images must be in a single folder for each class. The `for each class` part doesn't apply in this case.

File: decavision/model_testing/testing.py [hunk 3]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+            # print image and top predictions
+            ##top_pred = np.argsort(result)[::-1][:5] # GET PREDS > THRESHOLD (0.50) KEEP IT AS A USER ARGUMENT. 
+            top_pred = result > threshold
+            # Name of the true class.
+            cls_pred_name = np.array(categories)[top_pred]
+            cls_pred_perc = result[top_pred] * 100
+            cls_true = self.values[os.path.basename(image_path)]
+            if plot:
+                fig, ax = plt.subplots()
+                if self.model.name in ["Inception", "Xception"]:
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: This is the case for all models that are not EfficientNets. This is because the EfficientNets have the rescaling in the model so it's not done when loading the images. The rescaling is however done for all the other models so whatever you have to do for 'Inception' and 'Xception' should have to be done for the others. You can change this if for what I do in the __init__ to select the right models.

File: decavision/model_testing/testing.py [hunk 4]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+
+        # classification report
+        print("\nClassification report")
+        print(classification_report(y, y_hat, target_names=labels, digits=4))
+
+        # confusion matrix
+        print("\n Confusion matrix")
+        print(multilabel_confusion_matrix(y, y_hat))
+            
+    def create_movie(self, image_path):
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: If I understand correctly you need to run `classify_images` before running this to have the saved images? I think this method should be independent. So you should call the other method in it and have the path to images as input. Maybe keep the option to use premade images in case people have already called `classify_images`.

File: decavision/model_testing/testing.py [hunk 5]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+                    ax.imshow(image_tensor[0].astype('uint8'), interpolation='nearest')
+                xlabel = 'Prediction :\n'
+                for (x, y) in zip(cls_pred_name, cls_pred_perc):
+                    xlabel += '{0}, {1:.2f}%\n'.format(x, y)
+                ax.set_xlabel(xlabel)
+                ax.set_xticks([])
+                ax.set_yticks([]) 
+                plt.tight_layout() 
+                plt.show()
+                fig.savefig("predicated_images/" + os.path.basename(image_path))
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: I think this should be optional. It's not always necessary to save the results.

File: decavision/model_testing/testing.py [hunk 6]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+            path (str): location of the images
+            threshold (int): threshold for prediction (default to 0.5)
+        """
+    
+        # getting list of true and predicted labels
+        generator = self._load_dataset(path)
+        cls_true = generator.classes #true label for each image
+        labels = list(generator.class_indices.keys()) # list of total classes
+        cls_pred = self.model.predict(generator)
+        cls_pred = cls_pred > threshold
```
Comment (yangobeil) [inline decavision/model_testing/testing.py:418]: Isn't this already in the format that we want? And for the true labels they are also given by the generator in the right format. You just need to extract them and combine them in a single array. For example you could do something like
```
y_true = []
for i in range(generator.n):
    y_true.append(generator.next()[1])
y_true = np.array(y_true)
```

File: decavision/model_testing/testing.py [hunk 7]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+        print('Labels & Predictions loaded for reports')
+
+        # binarizer
+        mlb = MultiLabelBinarizer(classes = labels)
+        y = mlb.fit_transform(cls_true) 
+        y_hat = mlb.fit_transform(cls_pred_name)
+
+        # classification report
+        print("\nClassification report")
+        print(classification_report(y, y_hat, target_names=labels, digits=4))
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: The labels can be directly `self.categories`.

File: decavision/model_testing/testing.py [hunk 8]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+
+        # confusion matrix
+        print("\n Confusion matrix")
+        print(multilabel_confusion_matrix(y, y_hat))
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: I think you should specify the labels here too.

File: decavision/model_testing/testing.py [hunk 9]
```diff
@@ -253,3 +256,198 @@ def generate_classification_report(self, path):
...
+                                                y_col="labels",
+                                                target_size=self.input_shape,
+                                                shuffle=False,
+                                                color_mode='rgb',
+                                                class_mode='categorical',
+                                                classes=self.categories,
+                                                batch_size=1)    
+        return generator
+
+    def classify_images(self, path, categories, threshold = 0.5, plot=True):
```
Comment (yangobeil) [inline decavision/model_testing/testing.py]: You can use `self.categories` instead of asking them as input.

Comment (mitul23) [reply decavision/model_testing/testing.py]: done.

File: decavision/model_testing/testing.py (hunk 7 above)
Comment (mitul23) [reply decavision/model_testing/testing.py]: done.

File: decavision/model_testing/testing.py (hunk 2 above)
Comment (mitul23) [reply decavision/model_testing/testing.py]: done.

File: docs/multilabel_testing_example.rst [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+There are two ways to create a movie from classified images: you can directly run the following code by setting :code:`classify_images=True` to 
+make predictions on new images -> save classified images to a folder and create a movie from them:: 
+
+  tester.create_movie(path="image_dataset/val", classify_images=True, json_file=json_file)
+
+If you already have classified saved images in a folder, you can set :code:`classify_images=False` and pass an optional argument which will be the path 
+to the classified saved images directory :code:`**classified_path=classified_image_path`. Assume the classified images are saved under **image_dataset/classified_images/**, then::
+  
+  tester.create_movie(path=path, classify_images=False, json_file=json_file, classified_path="image_dataset/classified_images")
```
Comment (yangobeil) [inline docs/multilabel_testing_example.rst]: The argument `classified_path` changed name so you should change it here too.

**Review by yangobeil — APPROVED** [review]
Good job! This is a nice new functionality :)
